src/configuration/manager.ts

```diff
@@ -31,12 +31,30 @@ class ConfigurationManager implements IConfigurationManager {
 
     private readonly _fallback: IConfigurationFallbackMap;
 
-    constructor(fallback: IConfigurationFallbackMap) {
+    constructor(fallback: IConfigurationFallbackMap, deprecatedKeys: readonly string[]) {
         this._fallback = Object.assign<IConfigurationFallbackMap, IConfigurationFallbackMap>(Object.create(null), fallback);
+        this.showWarning(deprecatedKeys);
     }
 
     public dispose(): void { }
 
+    /**
+     * Shows an error message for each deprecated key, to help user migrate.
+     * This is async to avoid blocking instance creation.
+     */
+    private async showWarning(deprecatedKeys: readonly string[]): Promise<void> {
+        for (const key of deprecatedKeys) {
+            const value = vscode.workspace.getConfiguration("markdown.extension").get(key);
+            if (value !== undefined && value !== null) {
+                // We are not able to localize this string for now.
+                // Our NLS module needs to be configured before using, which is done in the extension entry point.
+                // This module may be directly or indirectly imported by the entry point.
+                // Thus, this module may be loaded before the NLS module is available.
+                vscode.window.showErrorMessage(`The setting 'markdown.extension.${key}' has been deprecated.`);
```

Comment:
This is too "aggressive", I mean everyone (millions) will see the pop-up.

I prefer to do a config transition in the background (without proactively notifying the users). (Of course we will say it in the changelog.)

e.g. https://github.com/James-Yu/LaTeX-Workshop/blob/8eb80f46794233c868223e99f5c206449bab5587/src/config.ts#L67

Reply:
> everyone (millions) will see the pop-up.

No.

https://github.com/yzhang-gh/vscode-markdown/blob/17c7bfc16361dd75dd6574b671fb37ac82e9ef44/package.json#L355-L359

I've set the default value of the deprecated setting to `null`. Thus, **only those that changed the setting** will see the notification.

---

> I prefer to do a config transition in the background

But **it's the user that knows the state of settings best**. A short message to the user can saves you tens of line of code.

As you can see from `KnownKey.ts`, `defaultFallback.ts`, and #888, the changes will be complex, involving **names, formats, and scopes**. You can **hardly** make it correct programmatically.

Reply:
> **only those that changed the setting** will see the notification.

Then this is good.

> But **it's the user that knows the state of settings best**.

From my experience most of the users know little about the settings... But for options like `syntaxDecorations` we can assume they know about it.
I was concerned someday we might change a configuration which affects majority of the users, although it isn't very likely to happen.

Reply:
> most of the users know little about the settings

I mean the settings are under the control of the user. **They remember what they did, and what those changes actually mean.**

If we tried to act on behalf of them, we would have to `inspect()` and analyze the possibly huge result. But it still cannot solve all problems. For example, in my plan, `decorationFileSizeLimit` (binary size) will be superseded by `maxDocumentLength` (the number of UTF-16 code units). I guess the factor is 1.5 (UTF-8, only BMP), but who knows, what if a user prefers Windows-1252 or Shift-JIS.

---

> someday we might change a configuration which affects majority of the users

Er ... Very likely.

I've identified a good number of flawed configuration keys, but only marked those that I've found a solution for as "To be superseded" in `KnownKey.ts`.

Reply:
>> most of the users know little about the settings
>
> I mean the settings are under the control of the user. **They remember what they did, and what those changes actually mean.**

Ideally, they should. The reality (from my observation) is, even many CS majoring students (my friends) have few concepts about the settings. It is just because they never pay attention to it.
I believe the fact "Markdown is rendered to HTML" will filter out 80% of the users. (Many users don't know what "decoration" means as it is a VS Code concept. And many many users never read the `README`.)
For average users, (probably) they just searched somewhere, changed the settings and totally forgot about it several months later. **They are not interested in how it should be configured**, they want _it just works_.

The "Windows-1252 and Shift-JIS" thing is a typical example. I would say more than 90% of the users don't know about it, as I have never seen it before 🤣.

Two of my own "rules":
- Only show pop-up message if we **really** want the users to know about it (e.g., exciting new features). And make sure the message is attractive (not like a regular non-exciting message)
- Do not assume the users {know well about / are interested in} how {Markdown / VS Code} works. ([curse of knowledge](https://en.wikipedia.org/wiki/Curse_of_knowledge))

Reply:
> Many users don't know what "decoration" means as it is a VS Code concept.

Do you think changing `theming` to `editorTheming` or `editing.theming` will help? E.g.

> `markdown.extension.editorTheming.decoration.renderCodeSpan`
>
> Markdown › Extension › Editor Theming › Decoration: Render Code Span

> `markdown.extension.editing.theming.decoration.renderCodeSpan`
>
> Markdown › Extension › Editing › Theming › Decoration: Render Code Span

---

> Windows-1252 and Shift-JIS

Windows-1252 was popular in Western Europe.
Shift-JIS is still common in Japan.

---

> curse of knowledge

OK. I'll keep it in mind when writing docs.

Reply:
> Do you think changing theming to `editorTheming` or `editing.theming` will help? E.g.

Ah, I didn't notice that. I like (just) `theming` or `theme` as it is very clear 👍, especially after `markdown` and `extension`. (It is much better than `syntax` that we had before)
The problem is at `decoration`, but I cannot think of a better one... Then I'd rather keep it aligned with VS Code (and "teach" users about it).

> Windows-1252 and Shift-JIS

Many thanks. #TodayILearned